barman/config.py

```diff
@@ -1314,6 +1323,234 @@ def _validate_with_keys(config_items, allowed_keys, section):
+        """
+        if set(obj.keys()) == set(cls._fields):
+            return cls(**obj)
+        if set(obj.keys()) == set(ConfigChange._fields):
+            return ConfigChange(**obj)
+        raise ValueError("Malformed configuration change serialization: %r" % obj)
+
+
+class ConfigChangesQueue:
+    """
+    Wraps the management of the config changes queue.
+
+    Once instantiated the queue can be accessed using the `queue` property.
+    """
+
+    def __init__(self, queue_file):
+        self.queue_file = queue_file
+        self._queue = None
+        self.open()
+
+    @staticmethod
+    def read_file(path) -> List[ConfigChangeSet]:
+        """Reads a json file containing a list of configuration changes."""
+        try:
+            with open(path, "r") as queue_file:
+                # Read the queue if exists
+                return json.load(queue_file, object_hook=ConfigChangeSet.from_dict)
+        except FileNotFoundError:
+            return []
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+    @property
+    def queue(self):
+        if self._queue is None:
+            self.open()
+        else:
+            return self._queue
+
+    def open(self):
+        self._queue = self.read_file(self.queue_file)
+
+    def close(self):
+        with open(self.queue_file + ".tmp", "w") as queue_file:
+            # Dump the configuration change list into the queue file
+            json.dump(self._queue, queue_file, cls=ConfigChangeSetEncoder, indent=2)
+
+        # Juggle with the queue files to ensure consistency of
+        # the queue even if Shelver is interrupted abruptly
+        old_file_name = self.queue_file + ".old"
+        try:
+            os.rename(self.queue_file, old_file_name)
+        except FileNotFoundError:
+            old_file_name = None
+        os.rename(self.queue_file + ".tmp", self.queue_file)
+        if old_file_name:
+            os.remove(old_file_name)
+        self._queue = None
+
+
+class ConfigChangesProcessor:
+    """
+    The class is responsible for processing the config changes to
+    apply to the barman config
+    """
+
+    def __init__(self, config):
+        self.config = config
+        self.applied_changes = []
+
+    def receive_config_changes(self, changes):
+        """
+        Process all the configuration changes
+        """
+        # Get all the available configuration change files in order
+        changes_list = []
+        for section in changes:
+            section_name = section["server_name"]
+            set = ConfigChangeSet(section_name)
+            for json_cng in section:
+                if json_cng in ("server_name", "scope"):
+                    continue
+                file_name = self.config._config.get_config_source(
+                    section_name, json_cng
+                )
+                if file_name == "default":
+                    file_name = "~/.barman.conf"
```

Comment:
```suggestion
                    file_name = os.path.expanduser("~/.barman.conf")
```

We need to perform the expansion through the `os` module, otherwise it tries to read/write a bogus file.